src/tribol/interface/tribol.hpp

```diff
@@ -313,23 +387,49 @@ int getCSRMatrix( int** I, int** J, real** vals, int csId,
  * \param [in]  csId Coupling scheme id
  * \param [in]  row_block Row Jacobian block (MORTAR, NONMORTAR, or LAGRANGE_MULTIPLIER)
  * \param [in]  col_block Column Jacobian block (MORTAR, NONMORTAR, or LAGRANGE_MULTIPLIER)
- * \param [out] row_elem_idx Reference to array of element indices for the row block
- * \param [out] col_elem_idx Reference to array of element indices for the column block
- * \param [out] jacobians Reference to array of Jacobian dense matrices
+ * \param [out] row_elem_idx Pointer to array of element indices for the row block
+ * \param [out] col_elem_idx Pointer to array of element indices for the column block
+ * \param [out] jacobians Pointer to array of Jacobian dense matrices
  *
  * \return 0 success (if Jacobians exist), nonzero for failure
  */
 int getElementBlockJacobians( integer csId, 
                               BlockSpace row_block,
                               BlockSpace col_block,
-                              axom::Array<integer>& row_elem_idx,
-                              axom::Array<integer>& col_elem_idx,
-                              axom::Array<mfem::DenseMatrix>& jacobians );
+                              const axom::Array<integer>* row_elem_idx,
+                              const axom::Array<integer>* col_elem_idx,
+                              const axom::Array<mfem::DenseMatrix>* jacobians );
+
+/*!
+ * @brief Get assembled contact contributions for the Jacobian matrix
+ *
+ * This method requires registration of an mfem::ParMesh with the coupling
+ * scheme. The Jacobian contributions are split into a 2x2 block structure. The
+ * first row (or column) block is associated with the coordinate grid function
+ * degrees of freedom and the second row (or column) block is associated with
+ * the Lagrange multiplier degrees of freedom.
+ *
+ * @param csId Coupling scheme id with a registered MFEM mesh
+ *
+ * @return Jacobian contributions as an mfem::BlockOperator
+ */
+std::unique_ptr<mfem::BlockOperator> getMfemBlockJacobian( integer csId );
+
+/// register mortar gaps scalar nodal field
+void registerMortarGaps( integer meshId,
+                         real * nodal_gaps );
 
-/// register a real nodal field
-void registerRealNodalField( integer meshId,
-                             const RealNodalFields field,
-                             real * fieldVariable );
+/// return mortar gaps on the non mortar surface for a given coupling scheme
+/// with a registered MFEM mesh
+mfem::ParGridFunction getMfemGap( integer cs_id );
+
+/// register mortar pressure scalar nodal field
+void registerMortarPressures( integer meshId,
+                              const real * nodal_pressures );
+
+/// return mortar pressures on the non mortar surface for a given coupling
+/// scheme with a registered MFEM mesh
+mfem::ParGridFunction& getMfemPressure( integer cs_id );
```

Comment:
Is there a reason this returns a reference and the 'getMfemGap()' and 'getMfemResponse()' don't?

Reply:
Yes, the gap vector and response vector are built on-the-fly (i.e. they are stored on the redecomp mesh and only transferred if these methods request them).  Pressure is stored on the submesh level in `MfemSubmeshData` allowing the host code to update pressure and have those updated values propogate down to the redecomp/Tribol level.

Note, the gap and response vector implementations have been updated to avoid a potential expensive copy (though mandatory copy elision in C++17 should fix this).